subt_ign/src/GameLogicPlugin.cc

```diff
@@ -1012,6 +1049,163 @@ void GameLogicPluginPrivate::OnEvent(const ignition::msgs::Pose &_msg)
+                      components::WorldLinearVelocity());
+                }
+
+                // Create an angular velocity component if one is not present.
+                if (!_ecm.Component<components::AngularVelocity>(link))
+                {
+                  _ecm.CreateComponent(link,
+                      components::AngularVelocity());
+                }
+
+                // Create a world angular velocity component if one is not
+                // present.
+                if (!_ecm.Component<components::WorldAngularVelocity>(
+                      link))
+                {
+                  _ecm.CreateComponent(link,
+                      components::WorldAngularVelocity());
+                }
+              }
+            }
+
+            if (mass > 0)
+            {
+              // This sets the kinetic energy threshold for the robot. It is
+              // based on the kinetic energy formula KE = 0.5 * M * v * v
+              //
+              // M is the mass of the robot.
+              // v is velocity of the robot. We are using the velocity acheived
+              // by falling from a height.
+              this->dataPtr->keInfo[model->Data()].kineticEnergyThreshold =
+                0.5 * mass * std::pow(
+                    sqrt((2 * this->dataPtr->keHeight) / 9.8) * 9.8, 2);
+              this->dataPtr->keInfo[model->Data()].robotName = mName->Data();
+
+              // Create a halt motion component if one is not
+              // present.
+              if (!_ecm.Component<components::HaltMotion>(model->Data()))
+              {
+                 _ecm.CreateComponent(model->Data(), components::HaltMotion());
+              }
+            }
+          }
+          return true;
+        });
+  }
+  else
+  {
+    // Check for crashes
+    for (auto &ke : this->dataPtr->keInfo)
+    {
+      ignition::gazebo::Link link(ke.second.link);
+      if (std::nullopt != link.WorldKineticEnergy(_ecm))
+      {
+        double currKineticEnergy = *link.WorldKineticEnergy(_ecm);
+
+        // We only care about positive values of this (the links looses energy)
+        double deltaKE = ke.second.prevKineticEnergy - currKineticEnergy;
+        ke.second.prevKineticEnergy = currKineticEnergy;
+
+        // Crash if past the threshold.
+        if (ke.second.kineticEnergyThreshold > 0 &&
+            deltaKE >= ke.second.kineticEnergyThreshold)
+        {
+          int64_t sec, nsec;
+          std::tie(sec, nsec) =
+            ignition::math::durationToSecNsec(_info.simTime);
+          ignition::msgs::Time localSimTime;
+          localSimTime.set_sec(sec);
+          localSimTime.set_nsec(nsec);
+
+          std::lock_guard<std::mutex> lock(this->dataPtr->eventCounterMutex);
+          // _resp.set_report_status("scoring finished");
+          std::ostringstream stream;
+          stream
+            << "- event:\n"
+            << "  id: " << this->dataPtr->eventCounter << "\n"
+            << "  type: kinetic\n"
```

Comment:
Can we use "collision" as the event name instead?

Reply:
Done in 56f62bcb12661f5e8338fa7f9abdadaa22775b59